### PR TITLE
rewrite bed compat code

### PR DESCRIPTION
**Pull request by BuckarooBanzay**
This PR rewrites the `beds` compatiibilty code to the new movenode-definition and callback system (see here: https://github.com/mt-mods/jumpdrive#movenode-compatibility and here: https://github.com/mt-mods/jumpdrive#hooks)
The new API removes lots of duplicate and expensive calls to `minetest.find_nodes_in_area` and should make jumping faster in general while also providing a generic way to listen to movement events.

I ended up rewriting larger portions of the code you made @tuedel (sorry :yum:)
Let me know if you have issues/suggestions with the way it is now

The new file is best viewed standalone and not as diff:
https://github.com/mt-mods/jumpdrive/blob/bed-compat-ng/compat/beds.lua

**NOTE**: this isn't tested yet, only a first draft

### CODE REVIEW

**Comment by tuedel** [comment]
Sure, go ahead :wink: My code was just quickly hacked together, I was planning on rewriting/simplifying it anyway.

**Comment by tuedel** [comment]
The sleeping position has changed in https://github.com/minetest/minetest_game/commit/9bf64b6, but just setting `y = y + 0.07` and using `vector.equals()` to didn't work for me because of floating point (im)precision.

I solved this on my server by just rounding down `y` for the check:
```
if sleep_pos.x == player_pos.x and
    math.floor(sleep_pos.y) == math.floor(player_pos.y) and
    sleep_pos.z == player_pos.z then

[...]
```
This conveniently also keeps backwards compatibility with the old sleeping positions.

File: compat/beds.lua [hunk 1]
```diff
@@ -1,5 +1,10 @@
 local bed_bottoms = {"beds:bed_bottom", "beds:fancy_bed_bottom"}
 
+-- sanity checks
+assert(beds)
+assert(beds.spawn)
+assert(beds.save_spawns)
```
Comment (Panquesito7) [inline compat/beds.lua:6]: Maybe show an error message if assert checks fail.

Comment (BuckarooBanzay) [reply compat/beds.lua:6]: > Maybe show an error message if assert checks fail.

I thought about that but letting it crash here ensures that the code does not execute any invalid stuff later, but you are right: a message would be nice, i think there is a second message parameter for the `assert()` command :thinking:

Comment (Panquesito7) [reply compat/beds.lua:6]: Yes, there is. It's a string value, you can print any much kind of messages.

**Comment by SwissalpS** [comment]
At this sate it seems to work.
What I was not able to test, is how players, currently in bed(s), are treated during a jump.
(Yes multiple players can be in multiple or the same one at the same time)

**Comment by BuckarooBanzay** [comment]
i'm fine with merging this if there are no obvious bugs (that result in crashes or corruption)

**Comment by SwissalpS** [comment]
after over four years ;) No more spawning without a bed... ... or spawning in the middle of space.

**Pull request merged**